Generic claim reconciler - example PR

p-strusiewiczsurmacki-mobica opened the pull request:
Example PR for review

schrej committed:
implement a generic claim reconciler in ipamutil

rikatz committed:
Prepare to use k8s image registry for releases

schrej commented on pkg/ipamutil/reconciler.go:
```diff
@@ -0,0 +1,225 @@
+
+	if !claim.ObjectMeta.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, claim)
+	}
```
Deletion is handled in line 146 as well. If we do it here like this, the address will never actually be released.

p-strusiewiczsurmacki-mobica replied:
Done

schrej commented on pkg/ipamutil/reconciler.go:
```diff
@@ -0,0 +1,225 @@
+		return r.reconcileDelete(ctx, claim)
+	}
+
+	// We always ensure there is a valid address object passed to the handler.
+	// The handler will complete it with the ip address.
+	if address.Name == "" {
+		address = newIPAddress(claim, handler.GetPool())
+	}
+
+	if res, err := handler.EnsureAddress(ctx, &address); err != nil || res != nil {
+		return unwrapResult(res), err
+	}
+
+	// Patch or create the address, ensuring necessary owner references are set.
+	operationResult, err := controllerutil.CreateOrPatch(ctx, r.Client, &address, func() error {
```
I think I made a mistake here. This would never cause the address to be patched, since the result of EnsureAddress would get overwritten if the IPAddress resource already exists. Instead, we should move lines 153-161 into the patch func passed to CreateOrPatch.

p-strusiewiczsurmacki-mobica replied:
Done

p-strusiewiczsurmacki-mobica replied:
It seems that this cannot be moved to the CreateOrPatch function:

```
E1018 11:57:33.140829   10185 controller.go:329] "Reconciler error" err="failed to create or patch address: MutateFn cannot mutate object name and/or object namespace" controller="ipaddressclaim" controllerGroup="ipam.cluster.x-k8s.io" controllerKind="IPAddressClaim" IPAddressClaim="test-ns-vdfmn/test" namespace="test-ns-vdfmn" name="test" reconcileID=508df50b-1d40-4ba1-bd15-aec888bef3a7
```

p-strusiewiczsurmacki-mobica replied:
I've just reverted that change for now.

schrej replied:
Ah, that error makes sense. The name and namespace need to be set before calling `CreateOrPatch`, otherwise it can't fetch the Address.
Can we create a new IPAddress that just has name and namespace set, and then set the references correctly during the mutateFn?

p-strusiewiczsurmacki-mobica replied:
Yes, I've just moved newIPAddress call outside of the mutateFn and it works OK now.

schrej commented on pkg/ipamutil/reconciler.go:
```diff
@@ -0,0 +1,225 @@
+}
+
+// ProviderIntegration is an interface that must be implemented by an IPAM provider.
+type ProviderIntegration interface {
```
I never liked this name. What do you think about calling it `ProviderAdapter` instead?

p-strusiewiczsurmacki-mobica replied:
Done

schrej commented on pkg/ipamutil/reconciler.go:
```diff
@@ -0,0 +1,228 @@
+		return ctrl.Result{}, nil
+	}
+
+	address := ipamv1.IPAddress{}
+	if err := r.Client.Get(ctx, req.NamespacedName, &address); err != nil && !apierrors.IsNotFound(err) {
+		return ctrl.Result{}, errors.Wrap(err, "failed to fetch address")
+	}
```
I think we don't need this anymore, the allocation of `address` can be moved down to line 150. The `CreateOrPatch` method will fetch the address for us, and `reconcileDelete` fetches it as well.

p-strusiewiczsurmacki-mobica replied:
Done

p-strusiewiczsurmacki-mobica committed:
Rebased and added changes required to pass all the tests

p-strusiewiczsurmacki-mobica committed:
Fixed linter issue

dependabot[bot] committed:
:seedling: Bump github.com/onsi/ginkgo/v2 from 2.13.0 to 2.13.1

Bumps [github.com/onsi/ginkgo/v2](https://github.com/onsi/ginkgo) from 2.13.0 to 2.13.1.
- [Release notes](https://github.com/onsi/ginkgo/releases)
- [Changelog](https://github.com/onsi/ginkgo/blob/master/CHANGELOG.md)
- [Commits](https://github.com/onsi/ginkgo/compare/v2.13.0...v2.13.1)

---
updated-dependencies:
- dependency-name: github.com/onsi/ginkgo/v2
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

k8s-ci-robot committed:
Merge pull request #166 from rikatz/build-using-k8s-infra

:seedling: Prepare to use k8s image registry for releases

schrej committed:
improve image building

support multi-arch images
align cloudbuild.yaml with cluster-api

k8s-ci-robot committed:
Merge pull request #205 from kubernetes-sigs/dependabot/go_modules/github.com/onsi/ginkgo/v2-2.13.1

:seedling: Bump github.com/onsi/ginkgo/v2 from 2.13.0 to 2.13.1

dependabot[bot] committed:
:seedling: Bump github.com/onsi/gomega from 1.29.0 to 1.30.0

Bumps [github.com/onsi/gomega](https://github.com/onsi/gomega) from 1.29.0 to 1.30.0.
- [Release notes](https://github.com/onsi/gomega/releases)
- [Changelog](https://github.com/onsi/gomega/blob/master/CHANGELOG.md)
- [Commits](https://github.com/onsi/gomega/compare/v1.29.0...v1.30.0)

---
updated-dependencies:
- dependency-name: github.com/onsi/gomega
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

k8s-ci-robot committed:
Merge pull request #204 from kubernetes-sigs/dependabot/go_modules/github.com/onsi/gomega-1.30.0

:seedling: Bump github.com/onsi/gomega from 1.29.0 to 1.30.0

dependabot[bot] committed:
:seedling: Bump github.com/onsi/ginkgo/v2 from 2.13.1 to 2.13.2

Bumps [github.com/onsi/ginkgo/v2](https://github.com/onsi/ginkgo) from 2.13.1 to 2.13.2.
- [Release notes](https://github.com/onsi/ginkgo/releases)
- [Changelog](https://github.com/onsi/ginkgo/blob/master/CHANGELOG.md)
- [Commits](https://github.com/onsi/ginkgo/compare/v2.13.1...v2.13.2)

---
updated-dependencies:
- dependency-name: github.com/onsi/ginkgo/v2
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

schrej committed:
naming and comments

schrej committed:
move Cluster watch to generic reconciler

schrej committed:
Merge pull request #208 from kubernetes-sigs/improve-image-building

🌱 improve image building

schrej committed:
remove image building github actions

k8s-ci-robot committed:
Merge pull request #209 from kubernetes-sigs/dependabot/go_modules/github.com/onsi/ginkgo/v2-2.13.2

:seedling: Bump github.com/onsi/ginkgo/v2 from 2.13.1 to 2.13.2

schrej committed:
disable arm image builds

it is unlikely that any relevant system will use arm and not arm64

schrej committed:
fix excludedAddresses property in readme

dependabot[bot] committed:
:seedling: Bump actions/setup-go from 4 to 5

Bumps [actions/setup-go](https://github.com/actions/setup-go) from 4 to 5.
- [Release notes](https://github.com/actions/setup-go/releases)
- [Commits](https://github.com/actions/setup-go/compare/v4...v5)

---
updated-dependencies:
- dependency-name: actions/setup-go
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

k8s-ci-robot committed:
Merge pull request #216 from kubernetes-sigs/dependabot/github_actions/actions/setup-go-5

:seedling: Bump actions/setup-go from 4 to 5

alexandrevilain committed:
🌱 Bump sigs.k8s.io/cluster-api from 1.4.1 to 1.6.0

schrej committed:
fix a few linter issues

schrej committed:
bump golangci-lint to 1.55.2

k8s-ci-robot committed:
Merge pull request #219 from alexandrevilain/feat/support-capi-1.6.0

🌱 Bump sigs.k8s.io/cluster-api from 1.4.1 to 1.6.0

p-strusiewiczsurmacki-mobica committed:
Incorporated code review issues

Signed-off-by: Patryk Strusiewicz-Surmacki <patryk-pawel.strusiewicz-surmacki@external.telekom.de>

p-strusiewiczsurmacki-mobica committed:
Fixed typo

Signed-off-by: Patryk Strusiewicz-Surmacki <patryk-pawel.strusiewicz-surmacki@external.telekom.de>

schrej commented:
I think we can close this one in favour of the one in the proper repo

schrej committed:
bump go version in ci to 1.21

schrej committed:
remove unused return parameter
